chains/lisk/watcher.go

```diff
@@ -0,0 +1,130 @@
+package lisk
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/golang/groupcache/lru"
+	"github.com/sisu-network/deyes/chains"
+	"github.com/sisu-network/deyes/chains/lisk/types"
+	"github.com/sisu-network/deyes/config"
+	"github.com/sisu-network/deyes/database"
+	"github.com/sisu-network/lib/log"
+)
+
+const (
+	TxTrackCacheSize = 1_000
+)
+
+func NewWatcher(db database.Database, cfg config.Chain,
+	client LiskClient) chains.Watcher {
+	blockCh := make(chan *types.Block)
+
+	w := &Watcher{
+		blockCh:      blockCh,
+		blockFetcher: newBlockFetcher(cfg, blockCh, client),
+		db:           db,
+		cfg:          cfg,
+		blockTime:    cfg.BlockTime,
+		client:       client,
+		lock:         &sync.RWMutex{},
+		txTrackCache: lru.New(TxTrackCacheSize),
+	}
+	return w
+}
+
+type Watcher struct {
+	cfg          config.Chain
+	client       LiskClient
+	blockTime    int
+	db           database.Database
+	vault        string
+	txTrackCache *lru.Cache
+	lock         *sync.RWMutex
+
+	// Block fetcher
+	blockCh      chan *types.Block
+	blockFetcher *defaultBlockFetcher
+}
+
+func (w *Watcher) init() {
+	vaults, err := w.db.GetVaults(w.cfg.Chain)
+	if err != nil {
+		panic(err)
+	}
+
+	if len(vaults) > 0 {
+		w.vault = vaults[0]
+		log.Infof("Saved gateway in the db for chain %s is %s", w.cfg.Chain, w.vault)
+	} else {
+		log.Infof("Vault for chain %s is not set yet", w.cfg.Chain)
+	}
+}
+
+func (w *Watcher) SetVault(addr string, token string) {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	log.Verbosef("Setting vault for chain %s with address %s", w.cfg.Chain, addr)
+	err := w.db.SetVault(w.cfg.Chain, addr, token)
+	if err == nil {
+		w.vault = strings.ToLower(addr)
+	} else {
+		log.Error("Failed to save gateway")
+	}
+}
+
+func (w *Watcher) Start() {
+	log.Info("Starting Watcher...")
+	w.init()
+	go w.scanBlocks()
+}
+
+func (w *Watcher) scanBlocks() {
+	go w.blockFetcher.start()
+	go w.waitForBlock()
+}
+
+// waitForBlock waits for new blocks from the block fetcher. It then filters interested txs and
+// passes that to receipt fetcher to fetch receipt.
+func (w *Watcher) waitForBlock() {
+	for {
+		block := <-w.blockCh
+		// Pass this block to the receipt fetcher
+		log.Info(w.cfg.Chain, " Block length = ", block.NumberOfTransactions)
+		for _, tx := range block.Transactions {
+			if strings.EqualFold(tx.Sender.Address, w.vault) {
+				log.Infof("Transfer %s from address %s to %s with message %s", tx.Asset.Amount, tx.Sender.Address, tx.Asset.Recipient.Address, tx.Asset.Data)
```

Comment:
You need to reconstruct the transaction and pass it back to the upper level through go channel.